client/components/OperationList.js

```diff
@@ -149,6 +148,115 @@ function ComputeAttachmentLink(op) {
     return `operations/${op.id}/`+file;
 }
 
+class LabelComponent extends React.Component {
+    constructor(props) {
+        has(props, 'operation');
+        has(props, 'buttonLabel');
+        super(props);
+        this.state = {
+            editMode: false            
+        };
+    }
+
+    dom() {
+        return this.refs.customlabel.getDOMNode();
+    }
+
+    switchToEditMode() {
+        this.setState({ editMode: true }, () => {
+            this.dom().focus();
+            //We set the cursor at the end
+            this.dom().selectionStart = (this.dom().value || '').length;
+        });
+    }
+    switchToStaticMode() {
+        this.setState({ editMode: false });
+    }
+
+    onBlur() {
+        let customLabel = this.dom().value;
+        Actions.SetCustomLabel(this.props.operation, customLabel);
+        // Be optimistic
+        this.props.operation.customLabel = customLabel;
+        this.switchToStaticMode();
+    }
+    
+    onKeyUp(e) {
+        if (e.keyCode == 13) {
+            this.onBlur();
+        }        
+    }
+    
+    defaultValue() {
+        if (this.props.operation.customLabel === null || this.props.operation.customLabel.trim().length === 0){
+            return this.props.label;
+        }
+        return this.props.operation.customLabel;
+    }
+    
+    render() {
+        if (!this.state.editMode){
+            return (
+                <button
+                  className="form-control text-left btn-transparent"
+                  id={this.props.operation.id}
+                  onClick={this.switchToEditMode.bind(this)}>
+                    {this.buttonLabel()}                    
+                </button>
+            );
+        }
+        return (
+            <input className="form-control"
+              type="text"
+              ref='customlabel'
+              id={this.props.operation.id}
+              defaultValue={this.defaultValue()}
+              onBlur={this.onBlur.bind(this)}
+              onKeyUp={this.onKeyUp.bind(this)}
+            />
+        );
+    }
+}
+
+class DetailedViewLabelComponent extends LabelComponent {
+    constructor(props) {
+        has(props, 'operation');
+        props.buttonLabel = () => {return this.buttonLabel()};
+        super(props);
+    }
+
+    buttonLabel() {
+        if (this.props.operation.customLabel === null ||  this.props.operation.customLabel.trim().length === 0){
+            return <em className="text-muted">{t('operations.add_custom_label') || "Add custom label"}</em>;
+        }
+        return <div className="label-button">{this.props.operation.customLabel}</div>;
+    }
+}
+
+class OperationListViewLabelComponent extends LabelComponent {
+    constructor(props) {
+        has(props, 'operation');
+        has(props, 'link');
+        has(props, 'label');
+        props.buttonLabel = (() => {return this.buttonLabel()});
```

Comment:
it seems props are read only (or modifications are deleted when calling the parent constructor). An 'elegant' solution could be: 
``` javascript
super({
operation: props.operation,
buttonLabel: (()=> { return this.buttonLabel()})});
```